lib/dd.ml

```diff
@@ -0,0 +1,2719 @@
+
+let make_literals () =
+  let res = Array.make (2 * _l_codes + 1) 0 in
+  res.(256) <- 1 ; res
+
+let succ_literal literals chr =
+  literals.(Char.code chr) <- literals.(Char.code chr) + 1
+let succ_length literals length =
+  assert (length >= 3 && length <= 255 + 3) ;
+  literals.(256 + 1 + _length.(length - 3)) <- literals.(256 + 1 + _length.(length - 3)) + 1
+
+let make_distances () = Array.make (2 * _d_codes + 1) 0
+
+let succ_distance distances distance =
+  assert (distance >= 1 && distance <= 32767 + 1) ;
+  distances.(_distance (pred distance)) <- distances.(_distance (pred distance)) + 1
+
+module N = struct
+  type dst = [ `Channel of out_channel | `Buffer of Buffer.t | `Manual ]
+
+  type dynamic =
+    { ltree : T.tree
+    ; dtree : T.tree
+    ; bltree : T.tree
+    ; h_lit : int
+    ; h_dst : int
+    ; h_len : int
+    ; symbols : int array }
+
+  let bl_tree ltree dtree ~bl_count =
+    let bl_freqs = Array.make (2 * _bl_codes + 1) 0 in
+    T.scan ltree.T.lengths ltree.T.max_code ~bl_freqs ;
+    T.scan dtree.T.lengths dtree.T.max_code ~bl_freqs ;
+
+    let bltree = T.make ~length:_bl_codes ~max_length:7 bl_freqs ~bl_count in
+    (* XXX(dinosaure): [T.make] needs [max_length] to avoid generation of a bad bltree (limited to 7 bits with extra). *)
+    let max_blindex = ref (_bl_codes - 1) in
+    let exception Break in
+
+    ( try while !max_blindex >= 3 do
+          if bltree.T.lengths.(zigzag.(!max_blindex)) <> 0
+          then raise_notrace Break ;
+          decr max_blindex
+        done
+      with Break -> () ) ;
+
+    !max_blindex, bltree
+
+  let dynamic_of_frequencies
+    : literals:int array -> distances:int array -> dynamic
+    = fun ~literals:lit_freqs ~distances:dst_freqs ->
+    let bl_count = Array.make (_max_bits + 1) 0 in
+    let ltree = T.make ~length:_l_codes lit_freqs ~bl_count in
+    let dtree = T.make ~length:_d_codes dst_freqs ~bl_count in
+    let max_blindex, bltree = bl_tree ltree dtree ~bl_count in
+    let bl_symbols = Array.make (_l_codes + _d_codes) 0 in
+    let i = T.symbols 0 ltree.T.lengths ltree.T.max_code ~bltree ~bl_symbols in
+    let i = T.symbols i dtree.T.lengths dtree.T.max_code ~bltree ~bl_symbols in
+    let bl_symbols = Array.sub bl_symbols 0 i in
+
+    { h_lit= ltree.T.max_code + 1
+    ; h_dst= dtree.T.max_code + 1
+    ; h_len= max_blindex + 1
+    ; bltree
+    ; ltree
+    ; dtree
+    ; symbols= bl_symbols }
+
+  let invalid_encode () = Fmt.invalid_arg "expected `Await encode"
+
+  type kind = Flat of int | Fixed | Dynamic of dynamic
+  type block = { kind: kind; last: bool; }
+
+  type encode = [ `Await | `Flush | `Block of block ]
+
+  let exists v block = match v, block.kind with
+    | (`Copy _ | `End), Flat _ ->
+      Fmt.invalid_arg "copy code in flat block can not exist"
+    | `Literal chr, Dynamic dynamic ->
+      dynamic.ltree.T.tree.Lookup.t.(Char.code chr) lsr _max_bits > 0
+    | `Copy (off, len), Dynamic dynamic ->
+      dynamic.l
```

Comment:
Are there legitimate cases where `len = 0` here?
Wonder if having an `assert (0 <> len);` would be worth it to catch decoder mistakes?

Reply:
If you talk about the `Flush` branch, `len` should never equal to `0`. An `assert` can be added here.